sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp

```diff
@@ -45,27 +44,25 @@ namespace Azure { namespace Core { namespace Http { namespace _detail {
     // Give access to private to this tests class
     friend class Azure::Core::Test::CurlConnectionPool_connectionPoolTest_Test;
 #endif
-  private:
-    // The cttor and dttor of this member makes sure of calling the libcurl global init and cleanup
-    AZ_CORE_DLLEXPORT static CurlGlobalStateForAzureSdk CurlGlobalState;
 
   public:
-    /**
-     * @brief Mutex for accessing connection pool for thread-safe reading and writing.
-     */
-    AZ_CORE_DLLEXPORT static std::mutex ConnectionPoolMutex;
-
-    /**
-     * @brief Keeps an unique key for each host and creates a connection pool for each key.
-     *
-     * @details This way getting a connection for a specific host can be done in O(1) instead of
-     * looping a single connection list to find the first connection for the required host.
-     *
-     * @remark There might be multiple connections for each host.
-     */
-    AZ_CORE_DLLEXPORT static std::
-        map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>>
-            ConnectionPoolIndex;
+    ~CurlConnectionPool()
+    {
+      using namespace Azure::Core::Http::_detail;
+      if (m_cleanThread.joinable())
+      {
+        {
+          std::unique_lock<std::mutex> lock(ConnectionPoolMutex);
+          // Remove all connections
+          g_curlConnectionPool.ResetPool();
+        }
+        // Signal clean thread to wake up
+        ConditionalVariableForCleanThread.notify_one();
+        // join thread
+        m_cleanThread.join();
+      }
+      curl_global_cleanup();
```

Comment:
Do we have to call this on behalf of the user? I thought the application calls init/cleanup on curl.

Reply:
Rick found the libcurl documentation which I never saw before where it is mentioned that curl_global_ functions uses a count ref, and libraries are expected to call the init and cleanup, which makes sure that if only the library is using libcurl, then only the library is calling them. And if the user is also using libcurl and calling the init/cleanup, then the calls from the library will be no-op

Reply:
Oh really?! Can you please share that doc link and even add it as a reference in the source as a comment.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/pull/1500

Reply:
https://curl.se/libcurl/c/libcurl.html

Reply:
A module like this must have global constant functions of its own, just like curl_global_init and curl_global_cleanup. The module thus has control at the beginning and end of the program and has a place to call the libcurl functions. Note that if multiple modules in the program use libcurl, they all will separately call the libcurl functions, and that's OK because only the first curl_global_init and the last curl_global_cleanup in a program change anything. (libcurl uses a reference count in static memory).

In a C++ module, it is common to deal with the global constant situation by defining a special class that represents the global constant environment of the module. A program always has exactly one object of the class, in static storage. That way, the program automatically calls the constructor of the object as the program starts up and the destructor as it terminates. As the author of this libcurl-using module, you can make the constructor call curl_global_init and the destructor call curl_global_cleanup and satisfy libcurl's requirements without your user having to think about it. (Caveat: If you are initializing libcurl from a Windows DLL you should not initialize it from DllMain or a static initializer because Windows holds the loader lock during that time and it could cause a deadlock.)